src/Algebra/Graph/Relation/Symmetric/Internal.hs

```diff
@@ -0,0 +1,193 @@
+-----------------------------------------------------------------------------
+-- |
+-- Module     : Algebra.Graph.Relation.Symmetric.Internal
+-- Copyright  : (c) Andrey Mokhov 2016-2019
+-- License    : MIT (see the file LICENSE)
+-- Maintainer : andrey.mokhov@gmail.com
+-- Stability  : unstable
+--
+-- This module exposes the implementation of derived binary relation data types.
+-- The API is unstable and unsafe, and is exposed only for documentation. You
+-- should use the non-internal module "Algebra.Graph.Relation.Symmetric" instead.
+-----------------------------------------------------------------------------
+
+module Algebra.Graph.Relation.Symmetric.Internal (
+    -- * Implementation of derived binary relations
+    SymmetricRelation (..), empty, vertex, overlay, connect, consistent,
+    referredToVertexSet, deduplicateSet, deduplicate
+  ) where
+
+import Control.DeepSeq (NFData (..))
+
+import Algebra.Graph.Relation.Internal hiding (empty, vertex, overlay, connect, referredToVertexSet, consistent)
+import Algebra.Graph.Relation (domain, relation, transpose)
+import qualified Data.Set as Set
+import Data.Monoid (mconcat)
+
+{-|  The 'SymmetricRelation' data type represents a /symmetric binary relation/
+over a set of elements. Symmetric relations satisfy all laws of the
+'Undirected' type class and, in addition, the
+commutativity of connect:
+
+@'connect' x y == 'connect' y x@
+
+The 'Show' instance produces symmetrically closed expressions:
+
+@show (1     :: SymmetricRelation Int) == "vertex 1"
+show (1 * 2 :: SymmetricRelation Int) == "edge 1 2"@
+
+The total order on graphs is defined using /size-lexicographic/ comparison:
+
+* Compare the number of vertices. In case of a tie, continue.
+* Compare the sets of vertices. In case of a tie, continue.
+* Compare the number of edges. In case of a tie, continue.
+* Compare the sets of edges.
+
+Here are a few examples:
+
+@'vertex' 1 < 'vertex' 2
+'vertex' 3 < 'Algebra.Graph.Relation.Symmetric.edge' 1 2
+'vertex' 1 < 'Algebra.Graph.Relation.Symmetric.edge' 1 1
+'Algebra.Graph.Relation.Symmetric.edge' 1 2 = 'Algebra.Graph.Relation.Symmetric.edge' 2 1
+'Algebra.Graph.Relation.Symmetric.edge' 1 1 < 'Algebra.Graph.Relation.Symmetric.edge' 1 2
+'Algebra.Graph.Relation.Symmetric.edge' 1 2 < 'Algebra.Graph.Relation.Symmetric.edge' 1 1 + 'Algebra.Graph.Relation.Symmetric.edge' 2 2
+'Algebra.Graph.Relation.Symmetric.edge' 1 2 < 'Algebra.Graph.Relation.Symmetric.edge' 1 3@
+
+Note that the resulting order refines the 'isSubgraphOf' relation and is
+compatible with 'overlay' and 'connect' operations:
+
+@'Algebra.Graph.Relation.Symmetric.isSubgraphOf' x y ==> x <= y@
+
+@'empty' <= x
+x     <= x + y
+x + y <= x * y@
+-}
+newtype SymmetricRelation a = SymmetricRelation { fromSymmetric :: Relation a }
```

Comment:
I think it may be better to name this data type simply `Relation`. It will always be possible to import the module qualified and then use `Symmetric.Relation` to avoid confusion, if need be. We follow this convention throughout the library, where we have `NonEmpty.Graph`, `Labelled.AdjacencyMap`, etc.

Reply:
Should I keep the constructor name?

Reply:
Better call the constructor `SR`, like we do with other data types in the library, to emphasise that this constructor is meant only for internal use.